// safety-lens/lib/Maybe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Just = function () {
  function Just(value) {
    _classCallCheck(this, Just);

    this.value = value;
  }

  _createClass(Just, [{
    key: 'map',
    value: function map(f) {
      return new Just(f(this.value));
    }
  }, {
    key: 'ap',
    value: function ap(x) {
      var f = this.value;
      if (x instanceof Just) {
        return new Just(f(x.value));
      } else {
        return x;
      }
    }
  }]);

  return Just;
}();

var Nothing = function () {
  function Nothing() {
    _classCallCheck(this, Nothing);
  }

  _createClass(Nothing, [{
    key: 'map',
    value: function map(f) {
      return this;
    }
  }]);

  return Nothing;
}();

var just = function just(val) {
  return new Just(val);
};
var nothing = new Nothing();

function traverseMaybe(f) {
  return function (pure, obj) {
    if (obj instanceof Just) {
      return f(pure, obj.value).map(just);
    } else {
      return pure(nothing);
    }
  };
}

exports.Just = Just;
exports.Nothing = Nothing;
exports.just = just;
exports.nothing = nothing;
exports.traverseMaybe = traverseMaybe;
